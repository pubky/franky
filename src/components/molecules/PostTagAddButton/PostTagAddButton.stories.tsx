import type { Meta, StoryObj } from '@storybook/nextjs-vite';
import { PostTagAddButton } from './PostTagAddButton';

const meta: Meta<typeof PostTagAddButton> = {
  title: 'Molecules/Post/PostTagAddButton',
  component: PostTagAddButton,
  parameters: {
    layout: 'centered',
    backgrounds: {
      default: 'dark',
      values: [
        { name: 'dark', value: '#05050A' },
        { name: 'light', value: '#F2F2F7' },
      ],
    },
    docs: {
      description: {
        component: 'Simple button with a plus icon to trigger adding a new tag. Features a dashed border.',
      },
    },
  },
  argTypes: {
    onClick: { action: 'clicked' },
  },
  tags: ['autodocs'],
};

export default meta;
type Story = StoryObj<typeof meta>;

export const Playground: Story = {
  render: (args) => <PostTagAddButton {...args} />,
};

export const Default: Story = {
  render: (args) => <PostTagAddButton {...args} />,
  parameters: {
    docs: {
      description: {
        story: 'Default add tag button with plus icon.',
      },
    },
  },
};

export const Interactive: Story = {
  args: {
    onClick: () => console.log('Add tag clicked'),
  },
  render: (args) => <PostTagAddButton {...args} />,
  parameters: {
    docs: {
      description: {
        story: 'Click the button to see the console log.',
      },
    },
  },
};

// With other tags
export const WithTags: Story = {
  render: (args) => (
    <div className="flex flex-wrap items-center gap-2">
      <div
        className="flex h-8 items-center gap-1.5 rounded-lg px-3 backdrop-blur-lg"
        style={{
          backgroundImage:
            'linear-gradient(90deg, rgba(5, 5, 10, 0.7) 0%, rgba(5, 5, 10, 0.7) 100%), linear-gradient(90deg, rgb(255, 153, 0) 0%, rgb(255, 153, 0) 100%)',
        }}
      >
        <span className="text-sm font-bold leading-5 text-white">bitcoin</span>
        <span className="text-sm font-medium leading-5 text-white opacity-50">21</span>
      </div>
      <div
        className="flex h-8 items-center gap-1.5 rounded-lg px-3 backdrop-blur-lg"
        style={{
          backgroundImage:
            'linear-gradient(90deg, rgba(5, 5, 10, 0.7) 0%, rgba(5, 5, 10, 0.7) 100%), linear-gradient(90deg, rgb(255, 0, 0) 0%, rgb(255, 0, 0) 100%)',
        }}
      >
        <span className="text-sm font-bold leading-5 text-white">based</span>
        <span className="text-sm font-medium leading-5 text-white opacity-50">7</span>
      </div>
      <PostTagAddButton {...args} />
    </div>
  ),
  parameters: {
    docs: {
      description: {
        story: 'Add button shown alongside existing tags.',
      },
    },
  },
};
