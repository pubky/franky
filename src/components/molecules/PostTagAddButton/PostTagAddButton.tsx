import * as React from 'react';
import * as Libs from '@/libs';
import * as Atoms from '@/atoms';

export interface PostTagAddButtonProps {
  /** Callback when button is clicked */
  onClick?: () => void;
  /** Additional className */
  className?: string;
}

export function PostTagAddButton({ onClick, className }: PostTagAddButtonProps) {
  return (
    <Atoms.Button
      onClick={onClick}
      variant="outline"
      size="sm"
      className={Libs.cn(
        'size-8 !min-w-8 p-0 rounded-lg border-dashed bg-transparent hover:bg-transparent transition-opacity hover:opacity-80',
        className,
      )}
      aria-label="Add new tag"
    >
      <Libs.Plus className="size-4 opacity-[0.32]" strokeWidth={2} />
    </Atoms.Button>
  );
}
