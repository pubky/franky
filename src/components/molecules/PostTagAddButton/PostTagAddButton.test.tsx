import { describe, it, expect, vi } from 'vitest';
import { render, screen, fireEvent } from '@testing-library/react';
import { PostTagAddButton } from './PostTagAddButton';

// Mock @/libs with partial mock
vi.mock('@/libs', async (importOriginal) => {
  const actual = await importOriginal<typeof import('@/libs')>();
  return {
    ...actual,
    Plus: ({ className, strokeWidth }: { className?: string; strokeWidth?: number }) => (
      <div data-testid="plus-icon" className={className} data-stroke-width={strokeWidth}>
        Plus
      </div>
    ),
  };
});

describe('PostTagAddButton', () => {
  it('renders button with plus icon', () => {
    render(<PostTagAddButton />);

    const button = screen.getByRole('button');
    const icon = screen.getByTestId('plus-icon');

    expect(button).toBeInTheDocument();
    expect(icon).toBeInTheDocument();
  });

  it('calls onClick when button is clicked', () => {
    const mockOnClick = vi.fn();
    render(<PostTagAddButton onClick={mockOnClick} />);

    const button = screen.getByRole('button');
    fireEvent.click(button);

    expect(mockOnClick).toHaveBeenCalledTimes(1);
  });

  it('has correct aria-label', () => {
    render(<PostTagAddButton />);

    const button = screen.getByLabelText('Add new tag');
    expect(button).toBeInTheDocument();
  });
});

describe('PostTagAddButton - Snapshots', () => {
  it('matches snapshot', () => {
    const { container } = render(<PostTagAddButton />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
