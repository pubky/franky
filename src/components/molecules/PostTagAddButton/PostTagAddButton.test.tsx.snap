// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PostTagAddButton - Snapshots > matches snapshot 1`] = `
<button
  aria-label="Add new tag"
  class="inline-flex items-center justify-center whitespace-nowrap text-sm disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*=\\"size-\\"])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/40 aria-invalid:border-destructive font-semibold cursor-pointer border shadow-xs hover:text-accent-foreground border-input gap-1.5 has-[>svg]:px-2.5 size-8 !min-w-8 p-0 rounded-lg border-dashed bg-transparent hover:bg-transparent transition-opacity hover:opacity-80"
  data-size="sm"
  data-slot="button"
  data-variant="outline"
>
  <div
    class="size-4 opacity-[0.32]"
    data-stroke-width="2"
    data-testid="plus-icon"
  >
    Plus
  </div>
</button>
`;
